src/Compilers/VisualBasic/Portable/Symbols/WellKnownMembers.vb

```diff
@@ -653,6 +653,75 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 Return MyBase.MatchTypeToTypeId(type, typeId)
             End Function
         End Class
+
+        Friend Function SynthesizeTupleNamesAttributeOpt(type As TypeSymbol) As SynthesizedAttributeData
+            Debug.Assert(type IsNot Nothing)
+            Debug.Assert(type.ContainsTuple())
+
+            Dim stringType = GetSpecialType(SpecialType.System_String)
+            Debug.Assert(stringType IsNot Nothing)
+            Dim names As ImmutableArray(Of TypedConstant) = TupleNamesEncoder.Encode(type, stringType)
+
+            ' If there are no names, elide the attribute entirely
+            If names.IsDefault Then
+                Return Nothing
+            End If
+
+            Dim stringArray = ArrayTypeSymbol.CreateSZArray(stringType, ImmutableArray(Of CustomModifier).Empty, stringType.ContainingAssembly)
+            Dim args = ImmutableArray.Create(New TypedConstant(stringArray, names))
+            Return TrySynthesizeAttribute(WellKnownMember.System_Runtime_CompilerServices_TupleElementNamesAttribute__ctorTransformNames, args)
+        End Function
+
+        Friend Class TupleNamesEncoder
+            Public Shared Function Encode(type As TypeSymbol) As ImmutableArray(Of String)
+                Dim namesBuilder = ArrayBuilder(Of String).GetInstance()
+
+                If Not TryGetNames(type, namesBuilder) Then
+                    namesBuilder.Free()
+                    Return Nothing
+                End If
+
+                Return namesBuilder.ToImmutableAndFree()
+            End Function
+
+            Public Shared Function Encode(type As TypeSymbol, stringType As TypeSymbol) As ImmutableArray(Of TypedConstant)
+                Dim namesBuilder = ArrayBuilder(Of String).GetInstance()
+
+                If Not TryGetNames(type, namesBuilder) Then
+                    namesBuilder.Free()
+                    Return Nothing
+                End If
+
+                Dim names = namesBuilder.SelectAsArray(Function(name, constantType)
+                                                           Return New TypedConstant(constantType, TypedConstantKind.Primitive, name)
+                                                       End Function,
+                                                       stringType)
+
+                namesBuilder.Free()
+                Return names
+            End Function
+
+            Private Shared Function TryGetNames(type As TypeSymbol, namesBuilder As ArrayBuilder(Of String)) As Boolean
+                type.VisitType(Function(t As TypeSymbol, builder As ArrayBuilder(Of String)) AddNames(t, builder), namesBuilder)
```

Comment:
I will keep the lambda here for consistency with C# and to be sure the delegate is cached.

---
In reply to: [79682730](https://github.com/dotnet/roslyn/pull/13923#discussion_r79682730) [](ancestors = 79682730)